ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1241 @@
+static int quic_write_nonblocking_epw(QUIC_CONNECTION *qc, const void *buf, size_t len,
+                                      size_t *written)
 {
+    /* Simple best effort operation. */
+    if (!ossl_quic_sstream_append(qc->stream0->sstream, buf, len, written)) {
+        /* Stream already finished or allocation error. */
+        *written = 0;
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+    }
+
+    quic_post_write(qc, *written > 0, 1);
     return 1;
 }
 
-QUIC_CONNECTION *ossl_quic_conn_from_ssl(SSL *ssl)
+int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
 {
-    return QUIC_CONNECTION_FROM_SSL(ssl);
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
+    int partial_write = ((qc->ssl_mode & SSL_MODE_ENABLE_PARTIAL_WRITE) != 0);
+
+    *written = 0;
+
+    if (!expect_quic_conn(qc))
+        return 0;
+
+    /* If we haven't started the handshake, do so automatically. */
+    if (!qc->started && !ossl_quic_do_handshake(qc))
+        return 0;
+
+    if (qc->ch != NULL && ossl_quic_channel_is_term_any(qc->ch))
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, SSL_R_PROTOCOL_IS_SHUTDOWN, NULL);
+
+    if (qc->stream0 == NULL || qc->stream0->sstream == NULL)
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+
+    if (blocking_mode(qc))
+        return quic_write_blocking(qc, buf, len, written);
+    else if (partial_write)
+        return quic_write_nonblocking_epw(qc, buf, len, written);
+    else
+        return quic_write_nonblocking_aon(qc, buf, len, written);
 }
 
 /*
- * The following are getters and setters of pointers, but they don't affect
- * the objects being pointed at.  They are CURRENTLY to be freed separately
- * by the caller the set them in the first place.
+ * SSL_read
+ * --------
  */
-int ossl_quic_conn_set_qrx(QUIC_CONNECTION *qc, OSSL_QRX *qrx)
+struct quic_read_again_args {
+    QUIC_CONNECTION *qc;
+    QUIC_STREAM     *stream;
+    void            *buf;
+    size_t          len;
+    size_t          *bytes_read;
+    int             peek;
+};
+
+static int quic_read_actual(QUIC_CONNECTION *qc,
+                            QUIC_STREAM *stream,
+                            void *buf, size_t buf_len,
+                            size_t *bytes_read,
+                            int peek)
 {
-    if (qc == NULL)
-        return 0;
-    qc->qrx = qrx;
+    int is_fin = 0;
+
+    if (stream->rstream == NULL)
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+
+    if (peek) {
+        if (!ossl_quic_rstream_peek(stream->rstream, buf, buf_len,
+                                    bytes_read, &is_fin))
+            return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+
+    } else {
+        if (!ossl_quic_rstream_read(stream->rstream, buf, buf_len,
+                                    bytes_read, &is_fin))
+            return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+    }
```

Comment:
Possible optimization as ossl_quic_rstream_peek() and read() are implemented as simple aliases to a single function might be to have only ossl_quic_rstream_read() with peek parameter which would indicate whether we want to just peek or also discard the read data.